relays/ethereum/src/ethereum_client.rs

```diff
@@ -363,3 +373,201 @@ impl EthereumHighLevelRpc for EthereumRpcClient {
 		Ok((id, transaction_receipts))
 	}
 }
+
+/// Substrate headers submitter API.
+#[async_trait]
+trait HeadersSubmitter {
+	/// Returns Ok(true) if not-yet-imported header is incomplete.
+	/// Returns Ok(false) if not-yet-imported header is complete.
+	///
+	/// Returns Err(()) if contract has rejected header. This probably means
+	/// that the header is already imported by the contract.
+	async fn is_header_incomplete(&self, header: &QueuedSubstrateHeader) -> Result<bool>;
+
+	/// Submit given header to Ethereum node.
+	async fn submit_header(&mut self, header: QueuedSubstrateHeader) -> Result<()>;
+}
+
+/// Implementation of Substrate headers submitter that sends headers to running Ethereum node.
+struct EthereumHeadersSubmitter {
+	client: EthereumRpcClient,
+	params: EthereumSigningParams,
+	contract_address: Address,
+	nonce: U256,
+}
+
+#[async_trait]
+impl HeadersSubmitter for EthereumHeadersSubmitter {
+	async fn is_header_incomplete(&self, header: &QueuedSubstrateHeader) -> Result<bool> {
+		let (encoded_call, call_decoder) =
+			bridge_contract::functions::is_incomplete_header::call(header.header().encode());
+		let call_request = CallRequest {
+			to: Some(self.contract_address),
+			data: Some(encoded_call.into()),
+			..Default::default()
+		};
+
+		let call_result = self.client.eth_call(call_request).await?;
+		let is_incomplete = call_decoder.decode(&call_result.0)?;
+
+		Ok(is_incomplete)
+	}
+
+	async fn submit_header(&mut self, header: QueuedSubstrateHeader) -> Result<()> {
+		let result = self
+			.client
+			.submit_ethereum_transaction(
+				&self.params,
+				Some(self.contract_address),
+				Some(self.nonce),
+				false,
+				bridge_contract::functions::import_header::encode_input(header.header().encode()),
+			)
+			.await;
+
+		if result.is_ok() {
+			self.nonce += U256::one();
+		}
+
+		result
+	}
+}
+
+/// Submit multiple Substrate headers.
+async fn submit_substrate_headers(
+	mut header_submitter: impl HeadersSubmitter,
+	headers: Vec<QueuedSubstrateHeader>,
+) -> SubmittedHeaders<SubstrateHeaderId, RpcError> {
+	let mut ids = headers.iter().map(|header| header.id()).collect::<VecDeque<_>>();
+	let mut submitted_headers = SubmittedHeaders::default();
+	for header in headers {
+		let id = ids.pop_front().expect("both collections have same size; qed");
+		submit_substrate_header(&mut header_submitter, &mut submitted_headers, id, header).await;
+
+		if submitted_headers.fatal_error.is_some() {
```

Comment:
Why not return an `Option<Error>` from `submit_substrate_header` instead of relying on the function to mutate passed arguments?